### PR TITLE
feat(support-for-error-templates): add the support of error templates

### LINKED ISSUE #52: Add a feature of optimized error message using error template
Error templates are custom-templated messages that can be provided for overriding default exception messages thrown in SDKs for error responses.

More information about this feature can be found [here](https://docs.apimatic.io/specification-extensions/swagger-codegen-extensions/#error-templates).

### PR DESCRIPTION
**Pull request by hamzashoukat94**
Exception from the server is significant. From server exception, we can take critical decisions. If the exception message from the server is not self-explanatory then it is hard for a client to respond accordingly.  We noticed that we can improve the message thrown to the client from the server.  Now, the client will register an error template against a particular code.  On throwing an exception, the core library will replace the placeholder using JsonPointer.  The user will get the proper message that he wanted.

closes #52

### CODE REVIEW

File: src/main/java/io/apimatic/core/ResponseHandler.java [hunk 1]
```diff
@@ -202,13 +204,28 @@ private void validateResponse(Context httpContext) throws ExceptionType {
...
         String errorCode = String.valueOf(statusCode);
+        String defaultErrorCode = "";
+        Matcher match = Pattern.compile("^[(4|5)[0-9]]{3}").matcher(errorCode);
+        if (match.find()) {
+            defaultErrorCode = errorCode.charAt(0) + "XX";
+        }
 
-        if (localErrorCases != null && localErrorCases.containsKey(errorCode)) {
-            localErrorCases.get(errorCode).throwException(httpContext);
+        if (localErrorCases != null) {
```
Comment (asadali214) [inline src/main/java/io/apimatic/core/ResponseHandler.java:213]: This piece of code is also repeating for globalErrorCases, cosider extracting it as a method, which return early it the hashMap is null

File: src/main/java/io/apimatic/core/ErrorCase.java [hunk 2]
```diff
@@ -41,7 +62,7 @@ private ErrorCase(final String reason, final ExceptionCreator<ExceptionType> exc
      * @throws ExceptionType Represents error response from the server.
      */
     public void throwException(Context httpContext) throws ExceptionType {
-        throw exceptionCreator.apply(reason, httpContext);
+        throw exceptionCreator.apply(getReason(httpContext), httpContext);
```
Comment (asadali214) [inline src/main/java/io/apimatic/core/ErrorCase.java:65]: this function should also hold the logic of getReason and replacePlaceHolder

File: src/main/java/io/apimatic/core/ResponseHandler.java (hunk 1 above)
Comment (hamzashoukat94) [reply src/main/java/io/apimatic/core/ResponseHandler.java]: Done

File: src/main/java/io/apimatic/core/ErrorCase.java (hunk 2 above)
Comment (hamzashoukat94) [reply src/main/java/io/apimatic/core/ErrorCase.java]: Done

File: src/test/java/apimatic/core/EndToEndTest.java [hunk 3]
```diff
@@ -99,6 +110,174 @@ public void testEndToEndSyncCall() throws IOException, CoreApiException {
...
     }
 
+
+    /**
+     * Test the local error template.
+     * @throws IOException Signals that an I/O exception of some sort has occurred.
+     * @throws CoreApiException Exception in api call execution.
+     */
+    @Test
+    public void testLocalErrorTemplateBody() throws IOException, CoreApiException {
```
[the remark this answers is not in the document]
Comment (hamzashoukat94) [reply src/test/java/apimatic/core/EndToEndTest.java:120]: Done

**Review by asadali214 — COMMENTED** [review]
These changes looks good to me

**Review by asadali214 — APPROVED** [review]
Approved

**Pull request merged**